stdlib/public/core/SmallString.swift

```diff
@@ -210,7 +204,10 @@ extension _SmallString {
   internal func withUTF8<Result>(
     _ f: (UnsafeBufferPointer<UInt8>) throws -> Result
   ) rethrows -> Result {
-    var raw = self.zeroTerminatedRawCodeUnits
+    // Code points need to be in little endian byte order and zero terminated.
+    let l = self.leadingRawBits
+    let t = self.trailingRawBits & _StringObject.Nibbles.largeAddressMask
+    var raw = (l.littleEndian, t.littleEndian)
```

Comment:
Why do we convert the bytes to little endian on read? If the plan is to store the contents in big-endian on big-endian systems, then we don't want to convert on read, right?

Reply:
And if not, then would it make more sense to just have a `littleEndianBytes` computed property?

Reply:
The trouble is that the discriminator is stored in the most significant byte of the `_object` pointer. On little endian systems this means the discriminator is stored in the 'last' byte of the string objects memory:
```
[ x x x x x x x x x x x x x x x d ]
```
However on big endian systems this means that the discriminator is stored in the middle byte of the string object:
```
[ x x x x x x x x d x x x x x x x ]
```
Since the discriminator is in the middle we can only access the first 8 code units with a `withUnsafeBytes` call if we use big endian byte order. However if we byte swap to little endian byte order before any `withUnsafeBytes` calls then we can access all 15 code units and so big endian machines can do all the same operations a little endian machine can do. It also means we can add code units to the small string in the same order on both big and little endian platforms which is much cleaner and easier to manage.

We could add a `littleEndianBytes` computed property but I don't think the code saving is worth the additional obfuscation. We should only ever need to do the little endian conversion before a `withUnsafeBytes` call and ideally we won't add any more direct calls to that and instead always go through `withUTF8` and `withMutableCapacity`.

Reply:
We could also go back to the approach I took in 2866b4a and always store `_SmallString._storage` in little endian byte order. That avoids byte swapping before a read but means that `_StringObject` and `_SmallString` have different raw values for the same string. Though this is also true for 32-bit architectures, so maybe we should go down that path. It would have the nice side effect of making `withUnsafeBytes` always safe on `_SmallString` values.

Reply:
Yeah, I think that's the path we'll want to go down. For 64-bit LE, `_StringObject` and `_SmallString` have equivalent layout, but otherwise we need to do extra unpacking / byte-swapping as needed.

That will probably make further maintenance easier as well, as we're more likely to add more readers than creators.

Reply:
... and might let any SIMD, which we really want to do (on LE platforms at the very least), at least have a chance of being portable.